Removed array from 2-sided binomtest and made computation O(logn) from O(n)

ryu577 opened the pull request:
For 2-sided version of binomial test, an array was unnecessarily created to find the index on the other side. This array creation was causing memory errors and was also very slow when large values of n were passed. Replacing this a simple binary search on an implicit array removed the additional memory footprint and also made the execution go from O(n) to O(logn)

<!-- 
Thanks for contributing a pull request! Please ensure that
your PR satisfies the checklist before submitting:
http://scipy.github.io/devdocs/dev/contributor/development_workflow.html#checklist-before-submitting-a-pr

Also, please name and describe your PR as you would write a
commit message:
https://docs.scipy.org/doc/numpy/dev/development_workflow.html#writing-the-commit-message

Note that we are a team of volunteers; we appreciate your
patience during the review process.

Again, thanks for contributing!
-->

#### Reference issue
<!--Example: Closes gh-WXYZ.-->

#### What does this implement/fix?
The two sided binomtest was O(n) memory and O(n) time. This was due to an array created for determining the index on the other side with binomial PMF terms. This array was unnecessary just for finding this index. In this PR, I eliminated the array and also replaced the search for the index from linear scan to binary search.

#### Additional information
<!--Any additional information you think is important.-->

mdhaber commented on scipy/stats/_binomtest.py:
```diff
@@ -313,16 +314,37 @@ def binomtest(k, n, p=0.5, alternative='two-sided'):
         elif k < p * n:
-            i = np.arange(np.ceil(p * n), n+1)
-            y = np.sum(binom.pmf(i, n, p) <= d*rerr, axis=0)
+            y = n-binary_search_for_binom_tst(a_fn,d*rerr,np.ceil(p * n),n)+1
```
```suggestion
            y = n-binary_search_for_binom_tst(a_fn, d*rerr, np.ceil(p * n), n)+1
```
PEP8 whitespace needed here and below

ryu577 replied:
Fixed with flake8 check.

mdhaber commented on scipy/stats/_binomtest.py:
```diff
@@ -300,7 +300,8 @@ def binomtest(k, n, p=0.5, alternative='two-sided'):
                          "must be 'two-sided', 'less' or 'greater'")
-
+    
+    a_fn = lambda x1:binom.pmf(x1,n,p)
```
```suggestion
    a_fn = lambda x1: binom.pmf(x1, n, p)
```
Does this need to be created up here or can it be done closer to where it's needed?
For instance, if it's going to be a `lambda` function, it could be passed into `binary_search_for_binom_tst` function directly.
If not, PEP8 suggests using `def` to create a named function. In that case, please use a more explicit name like `binom_pmf_np`.

ryu577 replied:
Moved inside the function call itself as suggested. PEP8 is okay with this.

mdhaber commented on scipy/stats/_binomtest.py:
```diff
@@ -313,16 +314,37 @@ def binomtest(k, n, p=0.5, alternative='two-sided'):
     return result
+
+
+def binary_search_for_binom_tst(a, d, lo, hi, asc_order=False):
```
```suggestion
def _binary_search_for_binom_tst(a, d, lo, hi, asc_order=False):
```
Names of private functions should begin with an underscore. Please include at least a brief docstring to indicate what the inputs are and what the function does.

Can standard library functions be used, e.g. from `bisect`?

At what size does this start becoming faster than the linear search? If it's quite large, please switch between the original linear approach and this at a reasonable value.

ryu577 replied:
bisect seems to work on arrays (https://docs.python.org/3/library/bisect.html), which defeats the purpose because the whole point here is to get rid of the array and binary search the function. Will do some tests on speed, but I suspect this will never be slower. We've been using this at my work place for quite a while (that usage was what caused me to discover the array search in the original function first place) and its always blazingly fast. Will add example usage to the file itself. Can you tell me where the test cases for this function are?

mdhaber replied:
Ok. How about scipy.optimize.bisect (or other bracket method in scipy.optimize.root_scalar)? I'm just trying to avoid adding a new implementation of a bisection algorithm as there's probably something suitable already.

ryu577 replied:
Yeah, I was looking for an existing library too. Couldn't find one that fit all needs. 

This function should return not exact match, but ix such that binom.pmf(ix)<d<binom.pmf(ix+1)

For scipy.optimize.bisect, it is meant to work on real valued functions and returns floats (not indices as is required here). The float will need to be converted to the appropriate index. If we think about all the edge cases, etc. this will become more complex than the binary search itself.

ryu577 replied:
Tried scipy.optimize.bisect, didn't return expected result.
>>> n=10; p=0.5
>>> a_fn = lambda x1:binom.pmf(x1,n,p)
>>> _binary_search_for_binom_tst(a_fn,0.1,0,np.floor(n*p),True)
## returns 2
>>>b_fn = lambda x1: a_fn(x1)-0.1
optimize.bisect(b_fn,0,np.floor(n*p))
## returns 4.999999999998863

ryu577 replied:
The reason this doesn't work is that binom.pmf(2,n,p) is 0.04394531249999999 and binom.pmf(3,n,p) is 0.11718750000000014. But for everything in between; ex: binom.pmf(2.2,n,p) you get 0. This throws all those root finding and bisection methods off. You could try to "fix" this behavior, but the code will become uglier than the binary_search.

ryu577 commented:
Can someone please let me know where the test cases for this function are specified? There are some I'd like to add.

mdhaber reviewed (changes requested):
Thanks; I think this will be useful. In addition to the specific comments here, please use `flake8` or similar to check this file for PEP8 compliance before the next commit.

mdhaber commented:
> Can someone please let me know where the test cases for this function are specified? There are some I'd like to add.

Looks like they're in `scipy/stats/tests/test_morestats.py`, Line 663.

ryu577 commented on scipy/stats/_binomtest.py:
```diff
@@ -327,7 +328,19 @@ def binomtest(k, n, p=0.5, alternative='two-sided'):
 
 
-def binary_search_for_binom_tst(a, d, lo, hi, asc_order=False):
+def _binary_search_for_binom_tst(a, d, lo, hi, asc_order=False):
```
@mdhaber - any update on this PR? I think I addressed all comments apart from the one about using an existing bisection function. I still don't see a good function that solves all requirements. My concern with using scipy.optimize.bisect is that it is meant for continuous functions. The binom.pmf function we need to do the binary search on is discrete, returning 0 for non-integer values. We could cast the values to integers, but this will make it a step function. Now, applying scipy.optimize.bisect on a step function can return any value in the range of the step. For ex: if fn(2)=0.1 and fn(3)=0.2 and we're searching for 0.15, it could return anything between 2 and 3 (in fact, in one such test it returns 3.000000000000682; what if tomorrow it starts returning 2.9999999958?). There is no way to always guarantee that it'll return either 2 or 3 (now and in all future versions of the function). And a change in behavior could break this function. Please advise - you still want to use it here?

mdhaber replied:
I thought that might be the trouble but i wanted to play with it to understand better. I haven't had time yet.
Without fully understanding, does that mean it c will get you within 1 of the correct integer? Is it possible to check the adjacent integer after it returns to see if that is the correct one?

mdhaber replied:
I played around with this. Something like this:
```python3
def _binary_search_for_binom_tst2(d, lo, hi):
    def f(i, d):
        i_low = np.minimum(np.floor(i), n-1)
        a, b = binom.pmf((i_low, i_low+1), n, p)
        c = a + (i - i_low) * (b - a)
        return c - d
    return root_scalar(f, bracket=(lo, hi), args=(d,)).root
```
would work. It just solves for the float `k` at which a linearly-interpolated binomial PMF would equal `d`, then you `np.floor` or `np.ceil` the result depending on the side you're looking at. Otherwise it works similarly to what you have, although I might have missed an edge case. It is usually a bit slower, but it also scales well.

I'm not sure which I prefer. If the one above were _faster_, I'd prefer it, but I suppose the default is to leave what you have alone. I think @WarrenWeckesser should take a look at this PR when he can, though, because he wrote `binomialtest`.

mdhaber replied:
Actually, we can stop when `d` is between `a` and `b`:
```python3
def _binary_search_for_binom_tst2(d, lo, hi):
    def f(k, d):
        k_low = np.minimum(np.floor(k), n-1)
        a, b = binom.pmf((k_low, k_low+1), n, p)
        return (0 if (a <= d <= b) or (a >= d >= b)
                else a + (k - k_low) * (b - a) - d)  # interpolate
    res = root_scalar(f, bracket=(lo, hi), args=(d,))
    return res.root
```
With that change, it begins to compete with `_binary_search_for_binom_tst`.

ryu577 replied:
I was going to try this over the next few days, but I'm glad you did :)
I tried your function and am having trouble:
>n=10; p=0.4
> d=0.22 #Its between pmf(3) and pmf(4)
>_binary_search_for_binom_tst2(d,0,5)

Return ValuError: f(a) and f(b) must have different signs (the original binary search version returns 3 with asc_order=True).

I expect this to be a bit slower since the root_scalar function doesn't know that the function is monotonically decreasing or monotonically increasing.

ryu577 replied:
I think it works well for the right side of the distribution but sometimes not for the left side. I'll try to fix it as well. Another fear I have (if I understand the function correctly) is that you set the function to 0 over the entire valid range that contains d (right?). In most cases, the root_scalar should return something in between that range. Then ceiling or floor will work. But, it could one day return the end points as well. For example, if pmf(3)<d<pmf(4); it could return anything between 3.0 and 4.0, inclusive (correct?).. then if you do ceil(3.2) you get 4 but if you do ceil(3.0), you get 3.0. This could be a rare edge case that never manifests though.

mdhaber replied:
I suppose we need to think about what we _want_ to happen when d equals one of the PMFs exactly (is it possible for values to the left and right of the mode to be equal?) and choose whether the inequality should be strict or not based on that.

mdhaber replied:
Do you know what the answer is? (Which index do we want to return when the PMF is equal on both sides?)

ryu577 replied:
If you look at the existing code, in the case of two-sided, they first calculate d, the pmf at input k value:

>d = binom.pmf(k,np)

Then they multiply d by some relative error:
> rerr = 1 + 1e-7

Then they look for the number of terms on the other side where the PMF <= d*rerr

> y = np.sum(binom.pmf(i, n, p) <= d*rerr, axis=0)

Not sure why they do this but with multiplying the rerr, but it effectively makes it a < condition instead of <=.

In real life, it'll have to be a perfect storm where d*rerr will happen to equal exactly a PMF on the other side (if it were just d without the multiplication with rerr, just a binomial PMF with p=0.5 would exercise this condition).

What I can say is that my code preserves existing behavior.

> n=10; p=0.5
> i=np.arange(np.ceil(p * n),n+1)
> y = np.sum(binom.pmf(i,n,p)<=binom.pmf(7,n,p),axis=0)

This returns 3. And my code:

> _binary_search_for_binom_tst(lambda x1:binom.pmf(x1,n,p),binom.pmf(7,n,p),0,np.floor(p*n),True)+1

Also returns 3.

WarrenWeckesser replied:
While @mdhaber's approach reuses an existing root finder, I think the implementation of `_binary_search_for_binom_tst` is simple enough that I'm OK with using it as it is.  It stays in the domain of integer indices, whereas using `root_scalar` requires a conversion to a continuous dependent variable, so it feels a bit more complicated than necessary.   Sure, it is actually pretty simple, and if using `root_scalar` turned out to be significantly faster, I'd say go with it.  But if it is borderline, then I'm fine with the existing bisection code.  Perhaps we can defer further investigation of optimizing the performance to a follow-up PR.

mdhaber replied:
Sounds good. I can't verify that the `_binary_search_for_binom_tst` code will do the right thing in all cases, but if someone else can, the rest of the PR looks good to me.

ryu577 replied:
@mdhaber - I think the tests I've added are pretty extensive and expose it to almost all conditions conceivable and ensure it maintains existing behavior (with the added benefits of speed and memory). I covered d being on the high side and search on low, d being on low, search on high, d being equal to input on the other side (added explicit test on binary search for this since the condition is very difficult to exercise in usage of binomtest) and monotonic PMF. I can't think of anything else but if someone else can, happy to add more tests to verify.

mdhaber replied:
Oops, I agree that the tests are very good. I believe it works very well, and I certainly can't think of counter-examples. I just meant to encourage someone else to take a close look.

mdhaber commented on scipy/stats/_binomtest.py:
```diff
@@ -313,16 +312,73 @@ def binomtest(k, n, p=0.5, alternative='two-sided'):
+def test_binomtest():
+    """
+    Some tests for binomtest. Should print all True when run.
+    """
```
Please add this to the test suite - again, they're in `scipy/stats/tests/test_morestats.py` starting at Line 663.
Ideally, you'd [`@parametrize`](https://docs.pytest.org/en/stable/parametrize.html) the test, and you should use `assert_allclose` to compare the values.

mdhaber commented on scipy/stats/_binomtest.py:
```diff
@@ -313,16 +312,73 @@ def binomtest(k, n, p=0.5, alternative='two-sided'):
+
+def _binary_search_for_binom_tst(a, d, lo, hi, asc_order=False):
+    """
+    Conducts an implicit binary search on a funcion specified by a.
+    args:
+        a: The function over which to perform binary search. Its values
+            for inputs lo and hi should be in ascending or descending
+            order as required by asc_order.
+        d: The value to search.
+        lo: The lower end of range to search.
+        hi: The higher end of the range to search.
+        asc_order: Weather the values returned by the function are in
+                   ascending or descending order.
```
Please follow the [standard docstring](https://numpydoc.readthedocs.io/en/latest/format.html) format.

ryu577 commented:
@mdhaber / @rgommers / @WarrenWeckesser - need some help on this last test. Looking at the logs, looks like pval of 1 is being passed in the comparison of assert_allclose for this case:
>res = stats.binomtest(10079999, 21000000, 0.48)
>assert_allclose(res.pvalue, 0.979042561004596, rtol=1e-8)
Not sure why this is happening only for the Linux version when I'm explicitly asking it to compare with 0.979042561004596.

ryu577 commented:
This is probably due to some numerical issues in that Linux version. Note this only happens when large inputs are passed. The numerical error is probably making the pvalue become 1. I can remove these tests with the large inputs and that will cause the tests to pass on this OS as well.

>BinomTestResult(k=10079999, n=21000000, alternative='two-sided', proportion_estimate=0.4799999523809524, pvalue=1.0)

Again, other OS's don't have this issue.

mdhaber commented:
@ryu577 You can xfail the test on 32-bit using a decorator. [Here's](https://github.com/scipy/scipy/blob/1ff8f62fb0ae45999b3f421eb71caed5ec7c3e4f/scipy/optimize/tests/test_linprog.py#L1765) an example.

ryu577 commented:
Yup, adding the decorator worked. I think that resolves all comments?

mdhaber commented:
> I think that resolves all comments?

Yes. I think @WarrenWeckesser will want to comment, though. We discussed trying to add tests (if they don't already exist) for edge cases, such as when the PMF value we're searching for is very close (equal, if possible?) on both sides of the mode, and for cases in which the PMF monotonically decreases (e.g. `binom(2, 0.1)`).

ryu577 reviewed (commented):
Already added a test for symmetric case (p=0.5). Added one for monotonic PMF as well. Also added a new set of tests for confirming that the old behavior is maintained when the PMF is symmetric and an exact matching d is passed. Good to know about the fisher_exact test.

mdhaber commented:
Thanks for adding these. I think they will be helpful when @WarrenWeckesser takes a look.

rlucas7 commented on scipy/stats/_binomtest.py:
```diff
@@ -313,16 +312,69 @@ def binomtest(k, n, p=0.5, alternative='two-sided'):
+    Parameters
+    ----------
+    a : function
+      The function over which to perform binary search. Its values
+      for inputs lo and hi should be in ascending or descending
+      order as required by `asc_order.
```
I'm confused by the requirement here, is this that `a(lo) <= a(mid) <= a(hi)` ***for all** values of `lo, mid, hi` in the domain of the binomial distribution or something else?

ryu577 replied:
The requirement is that a(lo) <= a(x) <= a(hi) for all lo<=x<=hi or a(lo)=>a(x)=>a(hi), depending on the boolean switch. Just like with binary search, you need a sorted array between the inputs you pass to it.

rlucas7 reviewed (commented):
@ryu577 I left a couple comments and a couple questions on approach.

mdhaber commented on scipy/stats/_binomtest.py:
```diff
@@ -313,16 +312,69 @@ def binomtest(k, n, p=0.5, alternative='two-sided'):
+      The index, i between lo and hi
+      such that a(i)<=d<a(i+1)
+    """
+    while lo < hi:
+        mid = (lo+hi)//2
+        midval = a(mid)
+        if midval < d:
+            if asc_order:
+                lo = mid+1
+            else:
+                hi = mid-1
+        elif midval > d:
+            if asc_order:
+                hi = mid-1
+            else:
+                lo = mid+1
+        else:
+            return mid
+    if a(lo) <= d:
+        return lo
+    else:
+        return lo-(asc_order-0.5)*2
```
> Is there any reason why we couldn't apply the function a() to all arguments first

The function `a` is the binomial pdf. The point of the PR is to reduce the number of those evaluations that need to be performed from O(n) to O(log n). More discussion about standard functions is [here](https://github.com/scipy/scipy/pull/13433#discussion_r563474301).

ryu577 replied:
Yes, it saves both time and space. Why do an approximation when you have an exact method that runs quite fast?

ryu577 commented:
@mdhaber / @rlucas7 - incorporated all the changes. The build failures on some Windows distros seem to have nothing to do with my changes:

C:\hostedtoolcache\windows\Python\3.8.7\x64\Lib\site-packages\scipy\integrate\_ivp\tests\test_ivp.py:263: in test_integration_sparse_difference

Any idea?

ryu577 commented on scipy/stats/_binomtest.py:
```diff
@@ -313,16 +312,69 @@ def binomtest(k, n, p=0.5, alternative='two-sided'):
+      The higher end of the range to search.
+    asc_order : bool
+      Whether the values returned by the function, `a are in
+      ascending or descending order.
```
Good suggestion, incorporated.

ryu577 commented on scipy/stats/_binomtest.py:
```diff
@@ -313,16 +312,69 @@ def binomtest(k, n, p=0.5, alternative='two-sided'):
+      such that a(i)<=d<a(i+1)
+    """
+    while lo < hi:
+        mid = (lo+hi)//2
```
Done.

rlucas7 commented on scipy/stats/_binomtest.py:
```diff
@@ -313,16 +312,59 @@ def binomtest(k, n, p=0.5, alternative='two-sided'):
+    Conducts an implicit binary search on a funcion specified by `a.
+
+    Meant to be used on the binomial PMF for the case of two-sided tests
+    to obtain the value on the other side of the mode where the tail prbability
```
```suggestion
    to obtain the value on the other side of the mode where the tail probability
```

I'm a bit confused on the inclusion of `on the other side of the mode` part, maybe I missed this part but isn't the calculation being done on the ***same*** side of the mode?

ryu577 replied:
No, for the two sided test, we want to find the point on the other side that is as extreme as the one we have on the current side. Hence, its the value as extreme as observation on other side of the binomial mode.

rlucas7 commented on scipy/stats/_binomtest.py:
```diff
@@ -313,16 +312,59 @@ def binomtest(k, n, p=0.5, alternative='two-sided'):
+
+    Meant to be used on the binomial PMF for the case of two-sided tests
+    to obtain the value on the other side of the mode where the tail prbability
+    should be computed. Because the values past the mode are always in
+    ascending order, this makes binary search applicable.
```
```suggestion
    should be computed. The values on either side of the mode are always in
    order making binary search is applicable.
```

rlucas7 commented on scipy/stats/_binomtest.py:
```diff
@@ -313,16 +312,59 @@ def binomtest(k, n, p=0.5, alternative='two-sided'):
+
+def _binary_search_for_binom_tst(a, d, lo, hi, asc_order=False):
+    r"""
+    Conducts an implicit binary search on a funcion specified by `a.
```
```suggestion
    Conducts an implicit binary search on a function specified by `a`.
```

rlucas7 commented on scipy/stats/_binomtest.py:
```diff
@@ -313,16 +312,59 @@ def binomtest(k, n, p=0.5, alternative='two-sided'):
     return result
+
+
+def _binary_search_for_binom_tst(a, d, lo, hi, asc_order=False):
```
```suggestion
def _binary_search_for_binom_tst(a, d, lo, hi):
```
IIUC the ascending order argument is no longer needed

ryu577 replied:
Good catch. Removed.

rlucas7 commented on scipy/stats/_binomtest.py:
```diff
@@ -313,16 +312,59 @@ def binomtest(k, n, p=0.5, alternative='two-sided'):
-            i = np.arange(np.floor(p*n) + 1)
-            y = np.sum(binom.pmf(i, n, p) <= d*rerr, axis=0)
+            y = _binary_search_for_binom_tst(lambda x1: binom.pmf(x1, n, p),
+                                             d*rerr, 0, np.floor(p * n) + 1,
+                                             True)+1
```
```suggestion
                                             d*rerr, 0, np.floor(p * n) + 1)+1
```
IIUC the ascending order argument is no longer needed

ryu577 replied:
Yes, good catch. Removed.

rlucas7 commented on scipy/stats/tests/test_morestats.py:
```diff
@@ -686,6 +686,70 @@ def binom_test_func(x, n=None, p=0.5, alternative='two-sided'):
+    # Expected results here are from old binom_test function.
+    # The alternative will be left unspecified which defaults
+    # it to two-sided.
+    @pytest.mark.xfail_on_32bit("The large inputs make these tests "
+                                "sensitive to machine epsilon level")
```
if these tests are still failing on the 32 bit platforms could we lower the values so they aren't so sensitive to the machine epsilon? 

Alternatively could we use a skipif pytest annotation that checks for 32 vs 64 bit via a sys call? 
(Skipping the test on 64 bit)

ryu577 replied:
The whole point is to check for large inputs. The existing function takes very long for this scale. Not sure how to use skipif. Any sample? Also, you meant skip on 32 bit, right?

rlucas7 replied:
There are multiple ways to check for this, take a look at the note here: https://docs.python.org/3/library/platform.html#platform.architecture
as one suggestion for a way to check. 

>Not sure how to use skipif. Any sample? 

Take a look at this one using skips on factorial function from `scipy.special`, 
https://github.com/scipy/scipy/blob/277dbab1612ae32f02038a6a0df7e061c927a5cc/scipy/special/tests/test_basic.py#L1831-L1840

>Also, you meant skip on 32 bit, right?

Yes, IIUC the tests should pass on 64 bit systems, correct?

ryu577 replied:
Ok. The current "@pytest.mark.xfail_on_32bit" does cause the tests to pass (since they're ignored on 32-bit). So, everything is working now. Do you think I should switch to @pytest.mark.skipif instead? And if so, would the decorator be something like: @pytest.mark.skipif(sys.maxsize > 2**32)?

tylerjereddy replied:
Will it use so much physical memory that you're concerned that it may occasionally segfault/error out on 32-bit systems (i.e., even letting it run and be a known failure could still lead to issues on occasion?).

ryu577 replied:
No, its just that 32 bit machines don't have enough precision for those kinds of numbers. So, the tests always fail since they default to a p-value of 0 or something, lacking the numerical precision.

ryu577 replied:
BTW, these kinds of large inputs take a while for the current version to run (even on 64 bit machines) while with the new code you get the same result almost instantly.

rlucas7 commented on scipy/stats/tests/test_morestats.py:
```diff
@@ -686,6 +686,70 @@ def binom_test_func(x, n=None, p=0.5, alternative='two-sided'):
+        y2 = stats._binary_search_for_binom_tst(lambda x1:
+                                                stats.binom.pmf(x1, n, p),
+                                                d,
+                                                0, np.floor(p * n) + 1,
+                                                True) + 1
```
```suggestion
                                                0, np.floor(p * n) + 1) + 1
```

rlucas7 commented on scipy/stats/_binomtest.py:
```diff
@@ -313,16 +312,59 @@ def binomtest(k, n, p=0.5, alternative='two-sided'):
+            hi = mid-1
+        else:
+            return mid
+    if a(lo) <= d:
+        return lo
+    else:
+        return lo-1
```
I'm not clear on why these are necessary, in the tests and code usage there are cases where a `+ 1` is added to the result of the binary search call. Would dropping this block eliminate the need for the `+ 1`s or are these needed for another reason?

ryu577 replied:
We need a(x)<d<a(x+1) on termination (x is returned value). I don't think there is a guarantee that when the while terminates, lo will necessarily satisfy this (lo could surpass the d in the loop). We could have a(lo)<d<a(lo+1) or a(lo-1)<d<a(lo). No?

ryu577 replied:
I just confirmed both conditions can be exercised depending on inputs. For example, when (9,21,0.48) is passed, the if condition is exercised. When (4,21,0.48) is passed on the other hand, the else condition is exercised.

rlucas7 reviewed (commented):
@ryu577 This PR looks pretty good from what I can see. I left a couple typo fixes in the docstrings and one or two other minor cleanups. I also had a question about the if/else block at the end of the search.

ryu577 commented:
@mdhaber / @rlucas7 / @WarrenWeckesser - thoughts on merging this? I think the tests I've added are quite extensive, ensuring parity of results with existing behavior in almost any scenario this is likely to be encountered in the wild. I believe I've also addressed all comments (?)

mdhaber commented:
I wrote a script to check the two-sided pvalues produced by this PR against those produced in master for `N` and `k` up to 1000 and for several values of `p`: 

- ten random values (`np.random.seed(0); np.random.rand(10)`)
- `k/n`
- `(n-k)/n`
- `0`
- `0.5`, and 
- `1`. 

It found 2894 minor discrepancies out of 7507485 comparisons, all in cases where `p==k/n` or `p==0.5`. The maximum difference between the two pvalues is `1.7880141811588146e-12`, and all are cases where this PR returns a `pvalue` of `1` whereas `master` gives a `pvalue` that is slightly less than 1. Here are a few examples to help figure out the difference.
```
binomtest(4, 7, 0.5)
binomtest(6, 11, 0.5)
binomtest(7, 13, 0.5)
binomtest(15, 44, 15/44)
binomtest(13, 46, 13/46)
binomtest(3, 47, 3/47)
binomtest(484, 967, 0.5)  # largest discrepancy found
```

@ryu577 thanks for your patience on this. I am not aware of a situation in which these discrepancies would be a problem in practice, but I think it's important to resolve them in order to maintain user confidence in our code. I'm not sure which is right - this PR, or master - either way, thanks for helping us track it down!

ryu577 commented:
@mdhaber - great catch! Thanks for the due diligence here. This was indeed an edge case where the y values from the binary search were crossing the boundaries. I added a check ensuring they remain within the boundaries. Once these checks on the new commit pass, can you please run the new version of the code through your 7 million tests once again?

ryu577 commented:
@mdhaber - thanks for catching this, appreciate your due diligence. All these cases were due to an edge case where the y value from binary search was crossing the bounds of the permissible range by 1. I added a check to keep it within the range. Can you please run your script again to see if any more discrepancies remain?

mdhaber commented:
Running. I don't want to post it as a gist because it's too sloppy to commit, but [here](https://www.dropbox.com/s/4ry5tsupqxe0l00/exhaustive_test.py?dl=0)'s the file if you're interested.

mdhaber reviewed (approved):
Approval is mostly based on tests above; I no longer find discrepancies between this and the old behavior for `n` up to 1000. I have taken a look at the code, but wrote the exhaustive tests because I don't trust myself to think through the edge cases. 

The `_binary_search_for_binom_tst` looks reasonable if the interval is right, but that is not immediately obvious to me. The [mode of the binomial distribution](https://en.wikipedia.org/wiki/Binomial_distribution#Mode) is not quite as simple as either the floor or ceil of `p*n`, and I haven't carefully thought through why the code still works as it is. I have a hard time sitting myself down to do that because I would prefer to start by finding the exact mode, or at least using `(n+1)*p` instead of `n*p` in the logic. A comment about this in the code would help.

It would be nice to get another approval before merging!

WarrenWeckesser commented:
It might be a few more days before I take a close look at this. My first impressions were good, though, so if another reviewer gives it the thumbs up, my lack a response should not be a blocker for merging sooner.

ryu577 commented:
@mdhaber - thanks for approving. For your question on the mode, I aimed to maintain existing behavior there. So, I used the same bounds as were already there in the existing code. Let me think about why those are used and add a comment if I figure something out. Otherwise, @WarrenWeckesser might be able to provide insight since you said he wrote the original test. 

In the meantime, I added more comments and fixed one more super rare edge case. Can you please run your checks one last time? I will figure out how to run the file you shared in the meantime so I can do it myself (or something similar).

mdhaber commented:
> I added more comments and fixed one more super rare edge case

Can you suggest what sort of test I could add that might have caught this? Does it only happen for larger `n` or for specific `p`s?

ryu577 commented:
This is actually the case where the d we calculate as pmf(k,n,p) happens to match exactly another on the other side of the mode. I had added a test case for this explicitly (not through binomtest), but had flipped the conditions in that case. To exercise it through the binomtest interface is tricky because we end up multiplying d by rerr (which is 1+1e-7). So, we would need a PMF function such that PMF(k,n,p)*(1+1e-7) = PMF(j,n,p) where j and k lie on different sides of the mode. To do this, the PMF values have to be very close to each other and that only happens when n is very large or p is very small. Let me see if I can construct such a case.

mdhaber commented:
Ok. Running the tests, and I'll post results in the morning.

mdhaber commented:
Everything still looks good.

ryu577 commented:
@WarrenWeckesser / @rlucas7 - any thoughts on taking a final look and giving me another signoff? I think the tests @mdhaber and I have put this through are very comprehensive at this point and I'm quite confident it preserves existing behavior in all situations while providing the speed and memory benefit. If you glance through the tests we've done both offline and added as official tests, you'll probably agree.

ryu577 commented:
@WarrenWeckesser / @rlucas7 - gentle ping :)

rlucas7 reviewed (changes requested):
@ryu577 I left one comment to clarify/answer your question and provided a link to a `skipif` example for the 64 bit platform issue. 

Other than that I think this PR is ready. 

Edit/update: after thinking this through a bit more we're achieving the same thing and the nuances of when to prefer `skipif` vs `xfail` don't seem worthwhile for change so let's leave this as is unless anyone else feels strong preference for skipif.

tylerjereddy commented on scipy/stats/tests/test_morestats.py:
```diff
@@ -686,6 +686,85 @@ def binom_test_func(x, n=None, p=0.5, alternative='two-sided'):
+        res = stats.binomtest(7, 13, 0.5)
+        assert_allclose(res.pvalue, 0.9999999999999999, rtol=1e-15)
+        res = stats.binomtest(6, 11, 0.5)
+        assert_allclose(res.pvalue, 0.9999999999999997, rtol=1e-15)
```
minor: this looks like a good place for parametrizing tests and expected results

but the again: queue various core devs saying it will slow the test suite down because then they become separate tests, so maybe not worth the turnover for now anyway

ryu577 replied:
Agree, I'd prefer to defer that to the next PR. But if you think I should do it in this one, please point me to some samples.

rlucas7 commented on scipy/stats/tests/test_morestats.py:
```diff
@@ -686,6 +686,85 @@ def binom_test_func(x, n=None, p=0.5, alternative='two-sided'):
+        res = stats.binomtest(10080017, 21000000, 0.48)
+        assert_allclose(res.pvalue, 0.9778567637538729, rtol=1e-15)
+
+    def tst_two_sided_pvalues2(self):
```
```suggestion
    def test_two_sided_pvalues2(self):
```

otherwise the testing harness does not pick up the tests.

rlucas7 commented on scipy/stats/tests/test_morestats.py:
```diff
@@ -686,6 +686,85 @@ def binom_test_func(x, n=None, p=0.5, alternative='two-sided'):
+        res = stats.binomtest(2, 4, .3)
+        assert_allclose(res.pvalue, 0.5883999999999999, rtol=1e-15)
+
+    def tst_edge_cases(self):
```
```suggestion
    def test_edge_cases(self):
```
similar to above

rlucas7 commented on scipy/stats/tests/test_morestats.py:
```diff
@@ -686,6 +686,85 @@ def binom_test_func(x, n=None, p=0.5, alternative='two-sided'):
+        res = stats.binomtest(6, 11, 0.5)
+        assert_allclose(res.pvalue, 0.9999999999999997, rtol=1e-15)
+
+    def tst_binary_srch_for_binom_tst(self):
```
```suggestion
    def test_binary_srch_for_binom_tst(self):
```

rlucas7 commented on scipy/stats/tests/test_morestats.py:
```diff
@@ -686,6 +686,85 @@ def binom_test_func(x, n=None, p=0.5, alternative='two-sided'):
+        # Old way of calculating y, probably consistent with R.
+        y1 = np.sum(stats.binom.pmf(i, n, p) <= d, axis=0)
+        # New way with binary search.
+        ix = stats._binary_search_for_binom_tst(lambda x1:
```
```suggestion
        ix = _binary_search_for_binom_tst(lambda x1:
```
function is private-not exported in `__all__`, which is fine-we wouldn't want. 

Also there needs to be a line:

```
from scipy.stats._binomtest import _binary_search_for_binom_tst
```
at the top of this file otherwise the function is not found, resulting in error.

rlucas7 commented on scipy/stats/tests/test_morestats.py:
```diff
@@ -686,6 +686,85 @@ def binom_test_func(x, n=None, p=0.5, alternative='two-sided'):
+        # Old way of calculating y.
+        y1 = np.sum(stats.binom.pmf(i, n, p) <= d, axis=0)
+        # New way with binary search.
+        ix = stats._binary_search_for_binom_tst(lambda x1:
```
```suggestion
        ix = _binary_search_for_binom_tst(lambda x1:
```
similar to above

rlucas7 reviewed (changes requested):
@ryu577 I left a couple small requests inline. The requested changes got it so the tests are passing for me locally on docker based ubuntu 64 bit.

The only request which isn't inline is to add: 

`from scipy.stats._binomtest import _binary_search_for_binom_tst`
at the top of  `scipy/stats/tests/test_morestats.py` file after removing `stats.` prefixes so that the function is available for use in the test file. 

Other couple ones are minor-and important-typos. 

Please let me know if anything is unclear.

ryu577 commented:
@rlucas7  - sorry for the delay. I think I addressed all your comments. Please let me know if anything remains missing.

ryu577 commented:
@rlucas7 / @mdhaber - not sure why the circleci and scipy tests are failing. Seem to have nothing to do with my changes. There are generally intermittent failures, but these have failed twice in a row. Any idea?

mdhaber commented:
You can ignore the circleci failure. I will submit an issue for that if there isn't one.
For the 32 bit failure, you use `@pytest.mark.xfail_on_32bit` there, too. I wouldn't be too bothered, personally. I can't see how your last two commits could have broken that.

rlucas7 commented on scipy/stats/tests/test_morestats.py:
```diff
@@ -686,6 +687,85 @@ def binom_test_func(x, n=None, p=0.5, alternative='two-sided'):
+
+    def test_two_sided_pvalues2(self):
+        res = stats.binomtest(9, n=21, p=0.48)
+        assert_allclose(res.pvalue, 0.6689672431938848, rtol=1e-15)
```
this one is failing too (similar to other failure in edge cases beneath), it passes for me on 64 bit ubuntu but not on the 32 bit CI:
https://dev.azure.com/scipy-org/SciPy/_build/results?buildId=10968&view=logs&j=6cac4766-9adb-5b92-83ee-b21ac03a0321&t=91c05f48-75b8-58d0-e0be-5f0ceaaac72f&l=14501
can you confirm this one too @ryu577 ? 

If the values are correct it looks like we could either:
(1) leave tests running on 32 & 64 bit and drop precision to `1e-11` or `1e-12` 
or 
(b) leave `rtol` settings and mark as xfail. 

My preference would be for the former, most pvalue calculations would not require such high precision and we'd have testing for both 32 and 64. 

I suppose option (3) might be to add a branch in test to set precision based on architecture (32 vs 64 get different precision) but might be more than is necessary here.

@WarrenWeckesser + @mdhaber thoughts/opinions?

ryu577 replied:
@rlucas7 - changed the precision to 1e-11 from 1e-15. Other tests seem to pass but circleci still fails.

mdhaber replied:
For circleci, you need to merge master. (Please merge rather than rebasing, since we're in the middle of the review.) Thanks.

mdhaber replied:
> My preference would be for the former... @WarrenWeckesser + @mdhaber thoughts/opinions?

Hah, actually I'd tend toward the latter and xfail only on 32 bit (`@pytest.mark.xfail_on_32bit`). Rationale is that if on 64 bit it's still achieving 1e-15, I can't imagine changes that would make it get worse just on 32 bit. However, if we only require 1e-11 on all platforms, its possible for 64-bit to change substantially without us noticing. (but maybe this isn't credible.) I don't know. It's probably fine either way. It would be really useful if `np.testing` could offer separate 32-bit tolerances.

rlucas7 replied:
> Hah, actually I'd tend toward the latter and xfail only on 32 bit (`@pytest.mark.xfail_on_32bit`). Rationale is that if on 64 bit it's still achieving 1e-15, I can't imagine changes that would make it get worse just on 32 bit.

Good point. Sorry to flip-flop here @ryu577 can you move tolerances back to `1e-15` and mark as xfail? 
 
> However, if we only require 1e-11 on all platforms, its possible for 64-bit to change substantially without us noticing (e.g. multiply all results by a factor of `1 + 1e-12` ... but maybe this isn't credible).

while this seems possible the changes seem specialized enough that it's unlikely we wouldn't notice changes that created this kind of behavior-I'm not worried about this sort of scenario.  

> I don't know. It's probably fine either way. It would be really useful if `np.testing` could offer separate 32-bit tolerances, or if there was some other easy way to do option 3.

I agree that option 3 might be best but let's get this PR merged w/the xfail markers. 
We can follow up with ideas to implement option 3 in future PR.

mdhaber replied:
Of course, just musing.

ryu577 replied:
@rlucas7  - reverted to 1e-15 and marked all tests as xfail on 32 bit.

rlucas7 reviewed (changes requested):
@ryu577 the two test failures look real I left a couple inline comments to take a look at-both solutions seem pretty straightforward changes (either adjust rtol or mark xfail). 
I indicated my preference inline but can see arguments for either.

Fixing up the changes for those two tests should get this to green.

rlucas7 commented:
@ryu577 thanks for the changes. It looks like things are green now except one. The only fail is a doc build from pythran which was fixed recently-but after you branched from master. Fix was via #13681 which was merged into master a few days ago. If you are comfortable doing so and want to rebase from the scipy master branch you can, otherwise I'll squash and merge on Sunday March 21st.

ryu577 commented:
@rlucas7 - I haven't done a rebase like that before, so I'll just wait for you to squash and merge on the 21st. In the meantime, any thoughts on giving me a sign-off?

rlucas7 commented:
> @rlucas7 - I haven't done a rebase like that before, so I'll just wait for you to squash and merge on the 21st. In the meantime, any thoughts on giving me a sign-off?

Ok no worries-I marked approved.

rlucas7 commented:
> > @rlucas7 - I haven't done a rebase like that before, so I'll just wait for you to squash and merge on the 21st. In the meantime, any thoughts on giving me a sign-off?
> 
> Ok no worries-I marked approved.

I didn't forget about this PR. There are some mypy issue on master when I merged another PR. Prefer to resolve that prior to merging this to avoid any confusion-it's a waiting game at this point.

rlucas7 commented:
squashed and merged-thanks @ryu577 and thanks @mdhaber @WarrenWeckesser and @tylerjereddy for the reviews.

ryu577 commented:
> squashed and merged-thanks @ryu577 and thanks @mdhaber @WarrenWeckesser and @tylerjereddy for the reviews.
> 
> Edit: I'll add a note to the release notes about the performance improvement.

Thanks a lot, @rlucas7 , @mdhaber , @WarrenWeckesser , @tylerjereddy . It was a great experience working on this PR with you folks. Hope to make more contributions to this great code base :)
